### PR TITLE
Add rewinder to handle enrichment off event poller critical path

### PR DESCRIPTION
**Pull request by peterbroadhurst**
Issue #663 demonstrated that the event loops of the various plugins cannot safely ask the DB for information the event aggregator loop is updating, concurrently to the event poller writing it.

However, we don't want to slow the event poller down by making it do complex enrichment tasks to work out how to rewind. These tasks are as follows:
- For blobs: Query all `data` IDs associated with a `blob` hash, then find all `message` IDs that are associated with that data
- For token transfers: Query the `batch` ID associated with the message, which is associated with the on-chain transfer

So this PR introduces the architecture summarized in the diagram below. This ensures we do **not** start the enrichment DB activities (in loop 2) until the event poller has completed the cycle that might have been active while the plugin event that generated the rewind was active

pr_665_diagram_1

### CODE REVIEW

File: internal/database/sqlcommon/message_sql.go [hunk 1]
```diff
@@ -498,6 +499,37 @@ func (s *SQLCommon) GetMessageIDs(ctx context.Context, filter database.Filter) (
 	return ids, nil
 }
 
+func (s *SQLCommon) GetMessagesBatchIDsForDataIDs(ctx context.Context, dataIDs []driver.Value) (batchIDs []*fftypes.UUID, err error) {
+	query := sq.Select("m.batch_id").From("messages_data AS md").LeftJoin("messages AS m ON m.id = md.message_id").Where(sq.Eq{"md.data_id": dataIDs})
```
Comment (awrichar) [inline internal/database/sqlcommon/message_sql.go:503]: JOIN 👀 👀 👀

File: internal/events/aggregator_rewind.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+func (rw *rewinder) start() {
+	go rw.rewindLoop1()
+	go rw.rewindLoop2()
+}
+
+// rewindLoop1 is responsible from taking requests to the aggregator for rewinds, from other
+// event loops that execute completely independently.
+// It takes them in as fast as possible, and puts them in a staging area ready for the aggregator
+// itself to mark it's finished processing anything in-flight
+func (rw *rewinder) rewindLoop1() {
```
Comment (awrichar) [inline internal/events/aggregator_rewind.go:89]: Would suggest more descriptive names than `rewindLoop1` and `rewindLoop2`... perhaps `receiveLoop` and `processLoop`?

File: internal/events/aggregator_rewind.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+		readyRewinds:     make(map[fftypes.UUID]bool),
+	}
+}
+
+func (rw *rewinder) start() {
+	go rw.rewindLoop1()
+	go rw.rewindLoop2()
+}
+
+// rewindLoop1 is responsible from taking requests to the aggregator for rewinds, from other
```
Comment (awrichar) [inline internal/events/aggregator_rewind.go:85]: _for_ taking requests

File: internal/events/aggregator_rewind.go [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+			// Pop when we've timed out
+			popCtx = timerCtx
+		} else {
+			// Only pop for new work, or shutdown
+			popCtx = rw.ctx
+		}
+		select {
+		case <-popCtx.Done():
+			if timerCtx == nil && timerCancel == nil {
+				log.L(rw.ctx).Debugf("Rewind loop 2 stopping")
```
Comment (awrichar) [inline internal/events/aggregator_rewind.go:129]: Should there just be separate `case` entries for `<-rw.ctx.Done()` and `<-timerCtx.Done()`? Seems like they're handled separately.

File: internal/events/aggregator_rewind.go [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+				timerCtx, timerCancel = context.WithTimeout(rw.ctx, rw.minRewindTimeout)
+			}
+		}
+	}
+}
+
+func (rw *rewinder) processStagedRewinds() bool {
+	batchIDs := make(map[fftypes.UUID]bool)
+	var msgRewinds []*fftypes.UUID
+	var newBlobHashes []driver.Value
```
Comment (awrichar) [inline internal/events/aggregator_rewind.go:153]: any reason not to type this to `*fftypes.Bytes32`?

File: pkg/database/plugin.go [hunk 6]
```diff
@@ -104,6 +105,12 @@ type iMessageCollection interface {
 
 	// GetMessagesForData - List messages where there is a data reference to the specified ID
 	GetMessagesForData(ctx context.Context, dataID *fftypes.UUID, filter Filter) (message []*fftypes.Message, res *FilterResult, err error)
+
+	// GetMessageBatchIDs - an optimized query to retrieve any non-null batch IDs for a list of message IDs
+	GetMessageBatchIDs(ctx context.Context, msgIDs []driver.Value) (batchIDs []*fftypes.UUID, err error)
```
Comment (awrichar) [inline pkg/database/plugin.go:110]: Can we not use `[]*fftypes.UUID` for the IDs?

File: internal/events/aggregator_test.go [hunk 7]
```diff
@@ -36,12 +35,14 @@ import (
...
 	"github.com/hyperledger/firefly/pkg/database"
 	"github.com/hyperledger/firefly/pkg/fftypes"
+	"github.com/sirupsen/logrus"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/mock"
 )
 
 func newTestAggregatorCommon(metrics bool) (*aggregator, func()) {
 	config.Reset()
+	logrus.SetLevel(logrus.DebugLevel)
```
Comment (awrichar) [inline internal/events/aggregator_test.go:45]: Is this needed here?

**Review by awrichar — APPROVED** [review]
Left some naming and typing suggestions, but functionally looks great. Nice to have all the rewind functionality consolidated now.

File: internal/database/sqlcommon/message_sql.go [hunk 8]
```diff
@@ -498,6 +499,37 @@ func (s *SQLCommon) GetMessageIDs(ctx context.Context, filter database.Filter) (
...
+	return s.queryBatchIDs(ctx, query)
+}
+
+func (s *SQLCommon) GetMessageBatchIDs(ctx context.Context, msgIDs []driver.Value) (batchIDs []*fftypes.UUID, err error) {
```
[the remark this answers is not in the document]
Comment (peterbroadhurst) [reply internal/database/sqlcommon/message_sql.go:507]: I'd put the `Message` first just to clarify why it's in the `message_sql` set of functions, but totally agree with the feedback. One tweak in what I changed it to:

- `GetBatchIDsForMessages` (as you suggested)
- `GetBatchIDsForDataAttachments` - added attachments to make it clear it's the messages

File: internal/events/aggregator_rewind.go (hunk 2 above)
Comment (peterbroadhurst) [reply internal/events/aggregator_rewind.go:89]: Thanks for the suggestions 👍  ... had scratched my head and left these names in out lack of inspiration

File: internal/events/aggregator_rewind.go (hunk 4 above)
Comment (peterbroadhurst) [reply internal/events/aggregator_rewind.go:129]: The problem is there isn't always a timer, so you get into complex logic where the select is dynamic. So this approach of having two options of which context to listen on, was the simplest thing I could find to do.

File: pkg/database/plugin.go (hunk 6 above)
Comment (peterbroadhurst) [reply pkg/database/plugin.go:110]: It means an extra loop and array allocation - I don't mind doing it here as it's a component boundary interface.

File: internal/events/aggregator_rewind.go (hunk 5 above)
Comment (peterbroadhurst) [reply internal/events/aggregator_rewind.go:153]: aha, it's valid in `sq.Eq`, but not in our builder `In("blob.hash", newHashes)` - so while I can remove the ones below on the `Database` package inputs, I can't remove this one without significant extra work.

File: pkg/database/plugin.go (hunk 6 above)
Comment (peterbroadhurst) [reply pkg/database/plugin.go:110]: (sorry comments here and above are a bit interleaved)

**Pull request merged**